crates/atiny-parser/src/io.rs

```diff
@@ -0,0 +1,222 @@
+        self.files.get(name).copied()
+    }
+}
+
+impl Dir {
+    pub fn add_node(&mut self, node_name: String, node: NodeId) {
+        self.files.insert(node_name, node);
+    }
+}
+#[derive(Clone)]
+pub struct File {
+    pub id: NodeId,
+    pub code: Rc<str>,
+    parent: Option<NodeId>,
+}
+
+pub enum Node {
+    File(File),
+    Dir(Dir),
+}
+
+impl Node {
+    pub fn parent(&self) -> Option<NodeId> {
+        match self {
+            Self::File(f) => f.parent,
+            Self::Dir(d) => d.parent,
+        }
+    }
+}
+
+pub type Reader<'r, T> = Box<dyn Fn(&T) -> Result<Rc<str>> + 'r>;
+
+pub struct VirtualFileSystem<'r, T: VirtualFile> {
+    root: String,
+    nodes: Vec<Node>,
+    reader: Reader<'r, T>,
+}
+
+impl<'r, T: VirtualFile> VirtualFileSystem<'r, T> {
+    pub fn new(path: &T, reader: impl Fn(&T) -> Result<Rc<str>> + 'r) -> Self {
+        let root = Node::Dir(Dir::new(NodeId(0), None));
+        Self {
+            root: path.dir_name(),
+            nodes: vec![root],
+            reader: Box::new(reader),
+        }
+    }
+
+    pub fn new_with(
+        path: &T,
+        reader: impl Fn(&T) -> Result<Rc<str>> + 'r,
+    ) -> Result<(Self, NodeId)> {
+        let mut vfs = Self::new(path, reader);
+        vfs.add_file(path, path.file_name(), None)
+            .map(|id| (vfs, id))
+    }
+
+    pub fn add_file(&mut self, path: &T, name: String, parent: Option<NodeId>) -> Result<NodeId> {
+        let file = File {
+            id: NodeId(self.nodes.len()),
+            parent,
+            code: (self.reader)(path)?,
+        };
+
+        self.add_node(Node::File(file), name, parent)
+    }
+
+    pub fn add_dir(&mut self, name: String, parent: Option<NodeId>) -> Result<NodeId> {
+        let dir = Dir::new(NodeId(self.nodes.len()), parent);
+        self.add_node(Node::Dir(dir), name, parent)
+    }
+
+    pub fn add_node(&mut self, node: Node, name: String, parent: Option<NodeId>) -> Result<NodeId> {
+        let id = NodeId(self.nodes.len());
+
+        let parent_node = match parent {
+            Some(NodeId(id)) => &mut self.nodes[id],
+            _ => &mut self.nodes[0],
+        };
+
+        let Node::Dir(dir) = parent_node else {
+            panic!("ICE: tried to add a node as a child of a file");
+        };
+
+        dir.add_node(name, id);
+        self.nodes.push(node);
+        Ok(id)
+    }
+
+    pub fn get_file_relative(&mut self, path: &str, NodeId(id): NodeId) -> Result<File> {
+        self.get_file_from_path(path, self.nodes[id].parent())
+    }
+
+    pub fn get_file_from_path(&mut self, path: &str, parent: Option<NodeId>) -> Result<File> {
+        let qualifier = path.split('.');
+
+        let mut dir = self.get_parent(parent);
+        let mut file = None;
+
+        for ty in qualifier {
+            if file.is_some() {
+                //todo: maybe support this in the future as inner modules
+                panic!("ICE: tried to treat a file as a directory")
```

Comment:
A panic handler would be good for this cases

Reply:
Agree with that